### ISSUE TITLE
How to train on a dataset of MR images?

### ISSUE THREAD

**Issue by hadign20** [oliverchampion/IVIMNET#3]
Hello,

I am trying to figure out a way to use IVIMNET with a dataset of MRI images.

I understand the `learn_IVIM` function takes an image:

``` python
def learn_IVIM(X_train, bvalues, arg, net=None):
    """
    This program builds a IVIM-NET network and trains it.
    :param X_train: 2D array of IVIM data we use for training. First axis are the voxels and second axis are the b-values
    :param bvalues: a 1D array with the b-values
    :param arg: an object with network design options, as explained in the publication check hyperparameters.py for
    options
    :param net: an optional input pre-trained network with initialized weights for e.g. transfer learning or warm start
    :return net: returns a trained network
    """
```

and splits the image voxels into train and val:

``` python
# splitting data into learning and validation set; subsequently initialising the Dataloaders
    split = int(np.floor(len(X_train) * arg.train_pars.split))
    train_set, val_set = torch.utils.data.random_split(torch.from_numpy(X_train.astype(np.float32)),
                                                       [split, len(X_train) - split])
    # train loader loads the trianing data. We want to shuffle to make sure data order is modified each epoch and different data is selected each epoch.
    trainloader = utils.DataLoader(train_set,
                                   batch_size=arg.train_pars.batch_size,
                                   shuffle=True,
                                   drop_last=True)
    # validation data is loaded here. By not shuffling, we make sure the same data is loaded for validation every time. We can use substantially more data per batch as we are not training.
    inferloader = utils.DataLoader(val_set,
                                   batch_size=min(val_set.__len__(),32 * arg.train_pars.batch_size),
                                   shuffle=False,
                                   drop_last=True)
```

My question is how would this work on a full dataset? Should we make a large 2D array from all images and feed the 2D array to the `learn_IVIM` function or should we change the structure of the code so that all the train/val split, batch normalization, etc happen on the images instead of voxels of one individual image?

Would appreciate any advice.

**Comment by oliverchampion** [comment]
Hey Hadign,

I think this is still a discussion point. If you see it as a fitting routine, I believe you can just use all your data to train and then using the same network, fit the data. As the loss is self supervised we haven't told the network what the ground truth is. In that case, you can put all your data into a 2d array and feed it to the train (1d voxels, 1d bvalues). This was the way we used it in our publications.

If you, on the other hand, want to train a network and use it on other data than the training data, you will need to do better bookkeeping. The provided train function will randomly split into train and Val, which is not done on a per patient base. Moreover, you will need to also separate a test set. But all in all, that are fairly easy adaptations. This would mean you have a trained network that others could in theory use. Nit also note that as the network is only trained for 1 value set, and means reusing a trained network may be less common due to that. Therefore, we use it as suggested in the first paragraph.

Cheers,

Oliver
